### ISSUE TITLE
Vocabulary inheritance through inclusion of base meta-schemas by references

### ISSUE THREAD

**Issue by gregsdennis** [json-schema-org/json-schema-spec#1153]
A [test](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/master/tests/draft2020-12/vocabulary.json) was added recently that checks to ensure that only keywords defined by included vocabs are evaluated.  It does this by declaring a [new meta-schema](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/master/remotes/draft2020-12/metaschema-no-validation.json) that doesn't have validation, then running a schema that includes a validation keyword.  The validator should skip this keyword.  I agree with the test, but it does open up a can of worms.

**Question:** Does a meta-schema inherit the vocabularies of the meta-schema(s) included by reference or only those vocabularies which are explicitly declared?

The fact that the root meta-schema for 2020-12 includes all of the vocabularies for the meta-schemas under it, even though all of the "child" meta-schemas each declare their own vocabulary sets, seems to imply that such inheritance doesn't take place and that a meta-schema must include all relevant vocabularies at the root level.

If the schema were changed to this

```json
{
  "$id": "http://localhost:1234/draft2020-12/metaschema-no-validation.json",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/applicator": true,
    "https://json-schema.org/draft/2020-12/vocab/core": true
  },
  "allOf": [
    { "$ref": "https://json-schema.org/draft/2020-12/schema" }  // this now points to root 2020-12
  ]
}
```

would the test still work, or does this meta-schema now include validation?

It seems like there's a subtlety here in that by referencing the full 2020-12 meta-schema, the keywords would be syntactically validated, but by not including the vocabularies which define those keywords, the logic for them shouldn't run.

---

A further nuance to this (that we should probably have a test around) is that we have decided that as a consequence of (or a necessity for) bundling, the meta-schema can change at any point, nested within the original schema.  How does this affect vocabularies?

```json
{
  "$schema": "http://localhost:1234/draft2020-12/metaschema-no-validation.json",
  "$defs": {
    "validateNormally": {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "const": "red"
    }
  },
  "$ref": "/$defs/validateNormally",
  "const": "green"
}
```

`"red"` would pass, naturally, because it's not checked for `metaschema-no-validation` and it's valid for `2020-12/schema`.  But what happens when I attempt to validate `"blue"` with this schema?

**Comment by jdesrosiers** [comment]
> Question: Does a meta-schema inherit the vocabularies of the meta-schema(s) included by reference or only those vocabularies which are explicitly declared?

No. `$vocabularies` only applies when it was included with `$schema`. That doesn't include schemas that are included from the meta-schema with `$ref`. I can't remember if the spec is actually clear about this or if I just know that was the intention from many discussions with Henry. I'd have to re-read that part of the spec, but I'm pretty sure it's in there somewhere. I'm sure it's explained differently than I did here tho.

> the meta-schema can change at any point, nested within the original schema.

The dialect can't change at _any_ sub-schema. A schema resource can only have one dialect, but the schema resources within a compound schema document can all have different dialects. A more direct way to put it is that `$schema` is only meaningful in a sub-schema if it has an `$id` sibling.

So, in your example, the `$schema` at `/$defs/validateNormally` has no effect and `"blue"` would pass validation. However, if there was an `$id` at that location as well, it would make that schema an embedded schema and a separate schema resource that uses the standard 2020-12 dialect. In this case, `"blue"` would fail validation on `"const": "red"`.

**Comment by Relequestual** [comment]
> > Question: Does a meta-schema inherit the vocabularies of the meta-schema(s) included by reference or only those vocabularies which are explicitly declared?
> 
> No. `$vocabularies` only applies when it was included with `$schema`. That doesn't include schemas that are included from the meta-schema with `$ref`. I can't remember if the spec is actually clear about this or if I just know that was the intention from many discussions with Henry. I'd have to re-read that part of the spec, but I'm pretty sure it's in there somewhere. I'm sure it's explained differently than I did here tho.

I concur. Vocabularies are only specified in meta-schemas. 
The root meta-schema for 2020-12 includes all of the vocabularies because it needs them at that root location due to the nature of dynamic references.

In a meta-schema (A), if a referenced schema (B) defines a different dialect, and it's meta-schemas (C) define a different set of vocabularies from the referring schema declaration, those defined set of vocabularies (In C) will be for the referenced schema (B) only.

Even if the referenced schema (B) uses a dynamic reference at its root to an anchor in the root of the initial meta-schema (A), because the initial meta-schema (A) defines its own dialect, that will be the one used applying the resulting referenced schema again.

In short, each schema resource (identified with `$id`) may identify the use of its own dialect, but they cannot make other schema resources use their defined dialect.

**Comment by karenetheridge** [comment]
> If the schema were changed to this [ referencing the full draft2020-12 metaschema in allOf, but without the $vocabulary declarations ] would the test still work, or does this meta-schema now include validation?

No it would not. The list of vocabularies that are used when evaluating a schema (each of which provide a list of keywords) is indicated by the `$vocabulary` keyword.  All the `allOf` is doing is adding extra syntax which is checked when a schema is treated as data and evaluated against the metaschema -- which is a little weird, but will work just fine.  You could add extra things in that `allOf` yourself if you wanted to restrict your metaschema in ways that don't relate to vocabularies -- for example, if you wanted to ensure that every schema had a `title` and `description`, or that all `$refs` were to locations in the same document.. or whatever else makes sense for your particular usecase.

It would also be perfectly valid (but weird) to include a bunch of vocabularies in the `$vocabulary` keyword and *not* reference the individual schemas themselves in the `allOf`. You're losing out on any syntax checking should you want to validate your schema against the metaschema, but technically your implementation should be doing that same validation itself as it parses the schema. (Whether it actually does in all cases is an implementation decision. In my case, most syntax checks that the schema would perform are actually duplicated in code, by necessity, e.g. so I ensure I don't dereference something that MUST be an array but actually is a syntactically-invalid string.. etc.)

Remember, all the "child metaschemas" in the `allOf` in the metaschema *aren't actually metaschemas themselves* - they're just extra schemas being included by a `$ref`.  See this part of the spec (https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.1.2): "The "$vocabulary" keyword MUST be ignored in schema documents that are not being processed as a meta-schema. This allows validating a meta-schema M against its own meta-schema M' without requiring the validator to understand the vocabularies declared by M."  It makes the evaluation of the `$vocabulary` keyword a bit convoluted, and there's a bit of recursion involved, but it all makes sense in the end.

> we have decided that as a consequence of (or a necessity for) bundling, the meta-schema can change at any point, nested within the original schema

The example you provided isn't valid because there is no `$id` adjacent to the `$schema` keyword: "The "$schema" keyword SHOULD be used in the document root schema object, and MAY be used in the root schema objects of embedded schema resources. It MUST NOT appear in non-resource root schema objects." (https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.1.1)  But otherwise...  I see no problem with switching metaschemas in the middle of evaluation. With bundling, this is a realistic usecase.  Whenever we're evaluating and we see a `$schema` keyword, we would need to switch the list of vocabularies to the new list and carry on... and then when done with that subschema, go back to the original list of vocabularies.  Because of the recursive nature of schema evaluation, implementing this turned out to be pretty easy - each level of subschema evaluation contains its own state, with settings inherited from parents and then overridden by keywords.

-------------

[aside: You did make me think of an important caveat to bundling, however... if the metaschema changes in the middle of the document and changes some important point of syntax, the inner schema might fail validation against some rules that the outer metaschema imparts.  I tried to think of an example but couldn't come up with anything that isn't really ugly though, and I think it could be fixed by careful use of the $dynamicRef.]

**Comment by gregsdennis** [comment]
> The "$vocabulary" keyword MUST be ignored in schema documents that are not being processed as a meta-schema.

This implies that if a meta-schema (being processed as such) doesn't contain a `$vocabulary` keyword, only `core` supplies (because it's always implied), and no other keywords should be evaluated.  Is that right?

This means that the root meta-schema alone defines what keywords in the schema should be evaluated.

**Comment by karenetheridge** [comment]
> This implies that if a meta-schema (being processed as such) doesn't contain a $vocabulary keyword, only core applies (because it's always implied), and no other keywords should be evaluated. Is that right?

My interpretation would be that it's invalid, because a metaschema must always have core.

> the root meta-schema alone defines what keywords in the schema should be evaluated

yes.

**Comment by gregsdennis** [comment]
https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.1.2.1 states

> If "$vocabulary" is absent, an implementation MAY determine behavior based on the meta-schema if it is recognized from the URI value of the referring schema's "$schema" keyword.

and

> If the implementation proceeds with processing the schema, it MUST assume the use of the core vocabulary.

This implies that a meta-schema without `$vocabulary` should implicitly include `core`, but can optionally refuse to process a schema that uses this meta-schema.

And (wow, it's right here) https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.1.2.2 states

> Note that the processing restrictions on "$vocabulary" mean that meta-schemas that reference other meta-schemas using "$ref" or similar keywords do not automatically inherit the vocabulary declarations of those other meta-schemas.

**Comment by gregsdennis** [comment]
So if `$vocabulary` specifies the set of keywords that should be processed, does `$schema` play any role in a meta-schema at all (aside from providing clues if `$vocabulary` is missing)?

**Comment by karenetheridge** [comment]
ok cool -- so it's reasonable to just default to the main metaschema for whatever draft we're using (which may be determined one layer earlier by a $schema keyword, or if totally absent, the implementation can do what it likes - my implementation defaults to the most recent draft, but that can be overridden with a config option).

`$schema` in a metaschema is for describing the schema itself (i.e. one layer back in the recursion). It would be unusual for a metaschema to use anything but one of the main draft metaschemas though -- but if it doesn't, something recognizable has to be back there eventually :)

**Comment by karenetheridge** [comment]
Yes, a schema can't be used as a metaschema for itself because it's not known yet, so how can it be used to validate itself when it doesn't know about itself yet?  The `$schema` keyword has similar functionality to a `$ref` in that sense -- first the document corresponding to that URI must be loaded in some way (or for implementations that don't support loading URIs from files/network, it must already be known internally.)

We can allow the "base" metaschemas to reference themselves, because we can bake their logic internally without having to actually load and parse the documents first.

**Comment by gregsdennis** [comment]
I got my answer for this.  Closing.
